File: PlantID/Features/Care/CareViewModel.swift
import SwiftData
import Observation
import Foundation

struct PlantCareItem: Identifiable {
    let plant: Plant
    let daysSinceWatering: Int
    let daysUntilDue: Int

    var id: UUID { plant.id }
    var isOverdue: Bool { daysUntilDue < 0 }
    var isDueToday: Bool { daysUntilDue == 0 }
    var isUpcoming: Bool { (1...2).contains(daysUntilDue) }
    var needsAttention: Bool { isOverdue || isDueToday }
}

@Observable
final class CareViewModel {
    var careItems: [PlantCareItem] = []
    var isLoading: Bool = true

    private let plantRepo: PlantRepository
    private let wateringLogRepo: WateringLogRepository

    init(plantRepo: PlantRepository, wateringLogRepo: WateringLogRepository) {
        self.plantRepo = plantRepo
        self.wateringLogRepo = wateringLogRepo
    }

    func loadData() {
        isLoading = true
        do {
            let plants = try plantRepo.getAlivePlants()
            careItems = computeItems(plants: plants)
        } catch {
            careItems = []
        }
        isLoading = false
    }

    func waterPlant(_ plantId: UUID) {
        guard let plant = careItems.first(where: { $0.id == plantId })?.plant else { return }
        let log = WateringLog(plant: plant)
        try? wateringLogRepo.insert(log)
        loadData()
    }

    private func computeItems(plants: [Plant]) -> [PlantCareItem] {
        plants.map { plant in
            let last = try? wateringLogRepo.getLastWatering(plantId: plant.id)
            let daysSince = last.map { Date().daysSince($0.wateredAt) } ?? -1
            let daysUntilDue = daysSince < 0
                ? -(plant.wateringIntervalDays + 1)
                : plant.wateringIntervalDays - daysSince
            return PlantCareItem(plant: plant, daysSinceWatering: daysSince, daysUntilDue: daysUntilDue)
        }
        .sorted { $0.daysUntilDue < $1.daysUntilDue }
    }
}

File: app/src/main/java/com/example/plant_id/ui/viewmodel/CareViewModel.kt
package com.example.plant_id.ui.viewmodel

import android.app.Application
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import com.example.plant_id.data.database.PlantDatabase
import com.example.plant_id.data.entity.Plant
import com.example.plant_id.data.entity.WateringLog
import com.example.plant_id.data.repository.PlantRepository
import com.example.plant_id.data.repository.WateringLogRepository
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch

data class PlantCareItem(
    val plant: Plant,
    val daysSinceWatering: Int,
    val daysUntilDue: Int
) {
    val isOverdue: Boolean get() = daysUntilDue < 0
    val isDueToday: Boolean get() = daysUntilDue == 0
    val isUpcoming: Boolean get() = daysUntilDue in 1..2
    val needsAttention: Boolean get() = isOverdue || isDueToday
}

class CareViewModel(
    private val plantRepository: PlantRepository,
    private val wateringLogRepository: WateringLogRepository,
    application: Application
) : AndroidViewModel(application) {

    var careItems by mutableStateOf<List<PlantCareItem>>(emptyList())
        private set

    var isLoading by mutableStateOf(true)
        private set

    init {
        loadData()
    }

    private fun loadData() {
        viewModelScope.launch {
            plantRepository.getAlivePlants().collect { plants ->
                computeCareItems(plants)
                isLoading = false
            }
        }
    }

    private suspend fun computeCareItems(plants: List<Plant>) {
        val now = System.currentTimeMillis()
        val items = plants.map { plant ->
            val last = wateringLogRepository.getLastWatering(plant.id)
            val daysSince = if (last != null) {
                ((now - last.wateredAt) / 86_400_000L).toInt().coerceAtLeast(0)
            } else {
                -1
            }
            val daysUntilDue = if (daysSince < 0) {
                -(plant.wateringIntervalDays + 1)
            } else {
                plant.wateringIntervalDays - daysSince
            }
            PlantCareItem(plant, daysSince, daysUntilDue)
        }
        careItems = items.sortedBy { it.daysUntilDue }
    }

    fun waterPlant(plantId: Long, onSuccess: () -> Unit = {}) {
        viewModelScope.launch {
            wateringLogRepository.insert(WateringLog(plantId = plantId))
            val plants = plantRepository.getAlivePlants().first()
            computeCareItems(plants)
            onSuccess()
        }
    }

    companion object {
        fun factory(application: Application): ViewModelProvider.Factory =
            object : ViewModelProvider.Factory {
                private val db = PlantDatabase.getInstance(application)
                private val plantRepo = PlantRepository(db.plantDao())
                private val wateringRepo = WateringLogRepository(db.wateringLogDao())
                @Suppress("UNCHECKED_CAST")
                override fun <T : ViewModel> create(modelClass: Class<T>): T =
                    CareViewModel(plantRepo, wateringRepo, application) as T
            }
    }
}
